### PR TITLE
Feat/Add support to blackbox-exporter to prometheus-exporter-operator

### PR DESCRIPTION
**Pull request by slopezz**
Closes https://github.com/3scale/platform/issues/456

This PR adds support for blackbox exporter to prometheus-exporter-operator.
- Added possibility to deploy a blackbox_exporter using  `PrometheusExporter` type `probe` (to monitor http endpoints, tcp endpoints... it depends on what you configure on the configmap)
- Added Grafana Dashboard to view stats of targets monitored by a specific blackbox-exporter
image

- Added full examples of what is needed to monitor different N targets:
  -  A single `PrometheusExporter` type `probe` object
  -  A single `Configmap` with the blackbox modules configuration (like http_2xx, tcp_connect...), that will be used by deployed `PrometheusExporter` type `probe` 
  - N `ServiceMonitor` objects that needs to point to the single deployed probe `PrometheusExporter` type `probe`,  which requires a specific relabeling of source_label `__param_target` into target_label `target`
- Adds documentation about the full provided examples of new `probe` exporter type
- Add examples of possible prometheus rules (alerts) using probe metrics (like status, or SSL Certificate expiration).

Taking into account that prometheus-operator had called `Probe` to the new custom resource that uses `blackbox_exporter`, it has been decided to use the same naming convention `PrometheusExporter` type `probe`  (details on https://github.com/prometheus-operator/prometheus-operator/pull/2832).

### CODE REVIEW

File: roles/prometheusexporter/templates/deployment.yml.j2 [hunk 1]
```diff
@@ -83,6 +83,14 @@ spec:
             - mountPath: /config/
               name: config-volume
               readOnly: true
+{% elif type == "probe" %}
+          args:
+            - --config.file=/etc/blackbox_exporter/config.yml
+            - --log.level=debug
```
Comment (roivaz) [inline roles/prometheusexporter/templates/deployment.yml.j2:89]: Shouldn't `debug` be a field in the Probe resource?

File: roles/prometheusexporter/templates/deployment.yml.j2 [hunk 2]
```diff
@@ -83,6 +83,14 @@ spec:
...
+          volumeMounts:
+            - mountPath: /etc/blackbox_exporter
+              name: config-volume
+              readOnly: true
```
Comment (roivaz) [inline roles/prometheusexporter/templates/deployment.yml.j2:93]: Does the blackbox-exporter binary watch for changes in the config file?

**Review by roivaz — COMMENTED** [review]
@slopezz wouldn't it be possible to add a field in the `Probe` spec to list the endpoints you want checked and have that automatically create the required ServiceMonitors? Seems a bit weird to me having to manually manage those.

File: roles/prometheusexporter/templates/deployment.yml.j2 (hunk 1 above)
Comment (slopezz) [reply roles/prometheusexporter/templates/deployment.yml.j2:89]: I harcoded it, because for blackbox_exporter, `debug` means showing that probes are being executed:
```
ts=2020-11-16T20:36:36.413Z caller=main.go:169 module=http_2xx target=https://kubernetes.io level=debug msg="Beginning probe" probe=http timeout_seconds=5 ts=2020-11-16T20:36:36.413Z caller=main.go:169 module=http_2xx target=https://kubernetes.io level=debug msg="Resolving target address" ip_protocol=ip4 
ts=2020-11-16T20:36:36.442Z caller=main.go:169 module=http_2xx target=https://kubernetes.io level=debug msg="Resolved target address" ip=147.75.40.148 ts=2020-11-16T20:36:36.442Z caller=main.go:169 module=http_2xx target=https://kubernetes.io level=debug msg="Making HTTP request" url=https://147.75.40.148 host=kubernetes.io 
ts=2020-11-16T20:36:36.758Z caller=main.go:169 module=http_2xx target=https://kubernetes.io level=debug msg="Received HTTP response" status_code=200 ts=2020-11-16T20:36:36.766Z caller=main.go:169 module=http_2xx target=https://kubernetes.io level=debug msg="Response timings for roundtrip" roundtrip=0 start=2020-11-16T20:36:36.442358373Z dnsDone=2020-11-16T20:36:36.442358373Z connectDone=2020-11-16T20:36:36.451026327Z gotConn=2020-11-16T20:36:36.460991651Z responseStart=2020-11-16T20:36:36.758412614Z end=2020-11-16T20:36:36.765988114Z 
ts=2020-11-16T20:36:36.766Z caller=main.go:169 module=http_2xx target=https://kubernetes.io level=debug msg="Probe succeeded" duration_seconds=0.352932335
```
So I can not imagine why would't you be interested in knowing that your configured probes on ServiceMonitors (which make use of your blackbox-exporter, on a separate object), are being executed (for me is like the expected output for `info` loglevel)

File: roles/prometheusexporter/templates/deployment.yml.j2 (hunk 2 above)
Comment (slopezz) [reply roles/prometheusexporter/templates/deployment.yml.j2:93]: Nope, any exporter is thought to be deployed on k8s, so you need to build that logic

Any of the exporters supports right know to roolout in case of secrets/configmaps.

On that specific case, the binary seems prepared https://github.com/prometheus/blackbox_exporter#configuration:
```
Blackbox exporter can reload its configuration file at runtime. If the new configuration is not well-formed, the changes will not be applied. A configuration reload is triggered by sending a SIGHUP to the Blackbox exporter process or by sending a HTTP POST request to the /-/reload endpoint.
```

Maybe it can be added a sidecar container to do so if it can be done on k8s, but actually, the exporter config is just used to define the modules, which are normally not changed over time (as the targets that will change more frequent)

File: roles/prometheusexporter/templates/deployment.yml.j2 (hunk 1 above)
Comment (raelga) [reply roles/prometheusexporter/templates/deployment.yml.j2:89]: Seems that's intencional: https://github.com/prometheus/blackbox_exporter/issues/399

**Review by raelga — APPROVED** [review]

File: roles/prometheusexporter/templates/deployment.yml.j2 (hunk 2 above)
Comment (raelga) [reply roles/prometheusexporter/templates/deployment.yml.j2:93]: This should be documented, maybe a commented disclaimer in the ConfigMap.

Comment (roivaz) [reply roles/prometheusexporter/templates/deployment.yml.j2:93]: You could always compute the hash of the config map and set it as a label in the Pod. When the ConfigMap changes, the label changes so a rollout is triggered. I don't know if you can calculate a hash in ansible, but maybe it's worth checking as it would be a very simple way of implementing config reload.

This could be extended to all of the other exporters, I think is an good feature that the operator could provide.

File: roles/prometheusexporter/templates/deployment.yml.j2 (hunk 1 above)
Comment (roivaz) [reply roles/prometheusexporter/templates/deployment.yml.j2:89]: Ok then, makes sense to have the logs in debug.

**Comment by slopezz** [comment]
>  @slopezz wouldn't it be possible to add a field in the Probe spec to list the endpoints you want checked and have that automatically create the required ServiceMonitors? Seems a bit weird to me having to manually manage those.

@roivaz  The purpose of the operator is deploying easily different prometheus exporters using simple `PrometheusExporter` resource (so you don't need to know that behind the scenes, there is a service, deployment with specific images, args, commands, secret/configmap mounts... and then a ServiceMonitor and its GrafanaDashboard). 

For this specfic case of blackbox, apart from deploying the exporter, in order to be able to use it because by its own does nothing, you need to create extra resources  for every endpoint to monitor:
- The idea would be to use the new `Probe` resource (where you need to specify the probe url of the deployed blackbox_exporter):
```yaml
apiVersion: monitoring.coreos.com/v1
kind: Probe
metadata:
  name: prometheus-exporter-probe-http-kubernetes-io
spec:
  prober:
    url: prometheus-exporter-probe-staging-blackbox:9115
  module: http_2xx
  targets:
    staticConfig:
      static:
        - https://kubernetes.io
```
- But taking into account this is a pretty new resource, so it is very possible it won't be available on your cluster, you can achieve the same result using the more standard `ServiceMonitor` resource. Instead of using using blackbox `prober.url` field, you need to use the `selector.matchLabels` pointing to the blackbox exporter service (so the same content on both cases). 
```yaml
apiVersion: monitoring.coreos.com/v1
kind: ServiceMonitor
metadata:
  name: prometheus-exporter-probe-http-kubernetes-io
spec:
  endpoints:
  - interval: 30s
    path: /probe
    port: metrics
    relabelings:
      - sourceLabels: [__param_target]
        targetLabel: target
    params:
      module:
      - http_2xx
      target:
      - https://kubernetes.io
  selector:
    matchLabels:
      app: prometheus-exporter-probe-staging-blackbox 
```
- In addition, the relabelling of the `target` label is intended for the dashboard only (you need a way to filter which probe is being used). Normally, when a metric comes from the exporter itself (like mysql, redis...), all source labels remains the same on prometheus  after being scraped by the `ServiceMonitor` (and you can use these metrics directly on grafana), however on that case, where you have an object (`Probe`/`ServiceMonitor`) pointing to the real deployed exporter (`prober.url`/`selector.matchLabels`), this target labels disappears, and is converted to `__param_target`, so in order to keep a label that identifies the probe, the relabelling is needed.

It would be nice to integrate an array of target/module into the blackbox `PromerheusExporter` resources (so you forget to create individual `Probes`/`ServiceMonitors`, but would be difficult to manage):
- Operator would require to create a new object with a unique name (important), and add the module/target values
   - name field should be required on the CR? maybe it could be used the hash of the target? it needs to be complian with kubernetes object naming (no dots, no special chars...)
- The problem comes when you delete targets from the array of targets, from ansible would be difficult to remove them, because if it disappear, on the next reconcile will ignore if it existed before and will be needed to be removed manually (don't know if I am explaining myself

So to summarize, I think by the moment, I would go with the standard prometheus probes monitoring, where you deploy an independent blackbox-exporter (with a `PromenetheusExporter` resource) and then create N resources for every target (`Probes` or `ServiceMonitors`) pointing to that deployed exporter, and then, the relabelling of the `target` label is intended for the dashboard only.

File: roles/prometheusexporter/templates/deployment.yml.j2 (hunk 2 above)
Comment (slopezz) [reply roles/prometheusexporter/templates/deployment.yml.j2:93]: Yes I will document it!

But I'm thinking that having the need of restarting the pod on different exporters due to secret/configmap changes, maybe it would be a good idea to add a sidecar container doing so

Comment (slopezz) [reply roles/prometheusexporter/templates/deployment.yml.j2:93]: By the moment I have just documented it on https://github.com/3scale/prometheus-exporter-operator/pull/13/commits/36d6a9fb3e19f829fc1132aeb54cb7bf638bbfa5 , a possible solution would be carried out on a new issue, because it affects to all exporters using Secrets/Configmaps (mysql, postgresql, cloudwatch, probe)

**Review by raelga — APPROVED** [review]

**Review by roivaz — APPROVED** [review]
